### ISSUE TITLE
pvactools may not identify missense variants corretly

### ISSUE THREAD

**Issue by worker000000** [griffithlab/pVACtools#611]
lastest docker pvactools; 
you can see there are a lot of missense variants here, but after running like following.
![image](https://user-images.githubusercontent.com/25199946/101163627-d43cf800-366e-11eb-9729-9e44f09e1f22.png)

![image](https://user-images.githubusercontent.com/25199946/101163394-832d0400-366e-11eb-85f0-64288c538baa.png)

2 why wecopy vep plugin by pvacseq. because the annotation done by vep, I just need to copy the plugin manullay by downloading the plugin first.  
I ran the following command by docker, but makes no change
docker run  -it --rm --user  `id -u`:`id -g` -v //:/data   griffithlab/pvactools pvacseqpvacseq install_vep_plugin <VEP plugins directory>

3 is the four steps needed to be done one after another
Annotating your VCF with VEP
Adding coverage data to your VCF
Adding expression data to your VCF
Creating a phased VCF of proximal variants

![image](https://user-images.githubusercontent.com/25199946/101165354-7bbb2a00-3671-11eb-8573-ed705b605f7e.png)
<br>
here it seems that snp and indel must be done separately？

4  /usr/bin/java -Xmx16g -jar /opt/GenomeAnalysisTK.jar \
-T ReadBackedPhasing \
-R reference.fa \
-I tumor.bam \
--variant combined_somatic_plus_germline.sorted.vcf \
-L combined_somatic_plus_germline.sorted.vcf \
-o phased.vcf

if I have matched sample, just I need to give the tumor.bam? how about the normal bam?

5  different caller gives different formats of vcf, varscan vardict mutect2 strekla2, how pvactools parse all correctly?

thanks a lot

**Comment by susannasiebert** [comment]
(1) Can you please delete the existing output directory and try again? It is finding an intermediate TSV file and using that instead of re-processing your VCF. That TSV might be corrupted. If you remove the existing output directory (or use a new output directory), it should process your VCF from scratch. If the problem persists you will need to share your VCF with us so that we can investigate.

(2) Yes, you will need to copy the VEP plugins before running VEP. The `pvacseq install_vep_plugin` command requires a destination path argument . In your example command I think this would be `/data`. So try `docker run -it --rm --user id -u:id -g -v //:/data griffithlab/pvactools pvacseqpvacseq install_vep_plugin /data`. Outside of your Docker container the files should then appear in `//`.

(3) Notice how the second command uses the output VCF from the first command as input. So the first command annotates only snvs and uses the `decomposed_vcf` (which has snvs and indels) to create a `snv_anntotated_vcf`. The `snv_anntotated_vcf` still has all of the snvs and indels in it, but only the snvs have readcounts added. The second command then takes this `snv_anntotated_vcf` and annotates only the indels (the snvs are outputted without any additional processing in this step). This creates the final `annotated_vcf`.  

@chrisamiller @jhundal @malachig or @obigriffith  would you be able to answer questions (4) and (5).

**Comment by chrisamiller** [comment]
4) read backed phasing is using information on reads from a single sample to infer haplotypes. Only one bam is needed.

5) Yes, merging calls from multiple callers can get complicated.  That is one reason why we often recommend using bam-readcount to add consistent VAF/depth information. You can see our process for merging and annotating variants in this portion of our CWL workflows: https://github.com/genome/analysis-workflows/blob/master/definitions/pipelines/detect_variants.cwl

Best of luck!

**Comment by 2236529177** [comment]
thanks a lot. it is really very discouraging.
after delete arguments  --vcf_info_field ANN and --gene_phenotype, the origin questions disapeared, 
but a more sorrowful bug comes, this is a bug you ever reported in vep . 
**Exception: Leading amino acid of the Downstream protein sequence (FLWFI) expected to match the wildtype amino acid at postion 39 (R). You may need to reannotate your VCF with a newer version of VEP**

<br>
after deleting the hgvs and hgvsg argument, another error comes out

**ERROR: There was a mismatch between the actual wildtype amino acid sequence (L) and the expected amino acid sequence (V). Did you use the same reference build version for VEP that you used for creating the VCF?
OrderedDict([('chromosome_name', '2'), ('start', '38298202'), ('stop', '38298203'), ('reference', 'C'), ('variant', 'G'), ('gene_name', 'CYP1B1'), ('transcript_name', 'NM_000104.3'), ('transcript_support_level', 'NA'), ('amino_acid_change', 'V/L'), ('codon_change', 'Gtg/Ctg'), ('ensembl_gene_id', '1545'), ('hgvsc', ''), ('hgvsp', ''), ('wildtype_amino_acid_sequence', 'MGTSLSPNDPWPLNPLSIQQTTLLLLLSVLATVHVGQRLLRQRRRQLRSAPPGPFAWPLIGNAAAVGQAAHLSFARLARRYGDVFQIRLGSCPIVVLNGERAIHQALVQQGSAFADRPAFASFRVVSGGRSMAFGHYSEHWKVQRRAAHSMMRNFFTRQPRSRQVLEGHVLSEARELVALLVRGSADGAFLDPRPLTVVAVANVMSAVCFGCRYSHDDPEFRELLSHNEEFGRTVGAGSLVDVMPWLQYFPNPVRTVFREFEQLNRNFSNFILDKFLRHCESLRPGAAPRDMMDAFILSAEKKAAGDSHGGGARLDLENVPATITDIFGASQDTLSTALQWLLLLFTRYPDVQTRVQAELDQVVGRDRLPCMGDQPNLPYVLAFLYEAMRFSSFVPVTIPHATTANTSVLGYHIPKDTVVFVNQWSVNHDPLKWPNPENFDPARFLDKDGLINKDLTSRVMIFSVGKRRCIGEELSKMQLFLFISILAHQCDFRANPNEPAKMNFSYGLTIKPKSFKVNVTLRESMELLDSAVQNLQAKETCQ'), ('downstream_amino_acid_sequence', ''), ('fusion_amino_acid_sequence', ''), ('variant_type', 'missense'), ('protein_position', '432'), ('transcript_expression', 'NA'), ('gene_expression', 'NA'), ('normal_depth', 'NA'), ('normal_vaf', 'NA'), ('tdna_depth', '1477'), ('tdna_vaf', '0.998645903859174'), ('trna_depth', 'NA'), ('trna_vaf', 'NA'), ('index', '129.CYP1B1.NM_000104.3.missense.432V/L'), ('protein_length_change', '')])**

<br>
it seems that because of reference genome. then I revised the argument  **--fasta** when run vep ,revised from Homo_sapiens.GRCh37.dna.primary_assembly.fa (vep release 99) to human_g1k_v37.fasta. it throws the same error.

about question 3
@susannasiebert 
I guesee if all these steps can be done in one command, it maybe better, or you can make it a shell script for tumor-only and matched samples, i t will be much better

@chrisamiller so do you mean, if I just give the vcf comes out from caller. such as mutect2 to vep , and then pvactools, it is not right.?  because AF in variant caller, especially gatk , **it is usually not just the devide alt and whole depth, it is something with bayes**. so I guess it maybe a little curtness to just use bam-readcount for all the caller.

thanks a lot

**Comment by 2236529177** [comment]
here I also have another question
1 /usr/bin/python /usr/bin/bam_readcount_helper.py  <decomposed_vcf> <sample_name> <reference_fasta> <bam_file> <output_dir>
for a matched tumor and normal sample, we have tumor and normal bam. but  this command will generate a snp and indel tsv file, so tumor bam generate two files and normal bam will generate two files?

so what should I do next?  do  bam-readcount -f <reference_fasta> -l <site_list> <bam_file> [-i] [-b 20]? or do vcf-readcount-annotator directly, if do vcf-readcount-annotator directly, it seems normal bam info ignored.

2 in no matter what step of question1, I seems to get a snv.vcf and indel.vcf, If I do not have rna expression data, I should go to phased, so I need to combine the vcf-readcount-annotator snp and indel vcf first?

thanks a lot

**Comment by susannasiebert** [comment]
> no matter what step of question1, I seems to get a snv.vcf and indel.vcf

This is incorrect. I think the naming of the VCFs in the example is just confusing you. You need to understand that every time you run the `vcf-readcount-annotator` you do so on the output VCF from when you ran the `vcf-readcount-annotator` the last time. It keeps on adding additional information to the same VCF.
```
vcf-readcount-annotator <input_vcf> <snv_bam_readcount_file> <DNA|RNA> \
-s <sample_name> -t snv -o <snv_annotated_vcf>
```
`snv_annotated_vcf` has exactly the same variants as `input_vcf`, all snvs and indels, but *only* the snvs will be annotated with readcounts. You now have a snv + indel VCF with only readcounts for snvs.
You now use `snv_annotated_vcf` as the *input* to the readcount step for indels:
```
vcf-readcount-annotator <snv_annotated_vcf> <indel_bam_readcount_file> <DNA|RNA> \
-s <sample_name> -t indel -o <annotated_vcf>
```
You now have `annotated_vcf` which has snvs *and* indels and both are annotated with readcounts. `annotated_vcf` is your only output VCF.

The above commands asks for a `sample_name`. If you use your tumor sample name, you will add readcounts to the tumor sample. You can use the same commands structure with your normal bams and the normal sample name to also add readcounts to the normal sample.
So first you would run the bam readcount helper like so:
```
/usr/bin/python /usr/bin/bam_readcount_helper.py <decomposed_vcf> <normal_sample_name> <reference_fasta> <normal_bam_file> <output_dir>

vcf-readcount-annotator <annotated_vcf> <normal_snv_bam_readcount_file> <DNA|RNA> \
-s <normal_sample_name> -t snv -o <annotated_vcf_plus_normal_snv_readcounts>

vcf-readcount-annotator <annotated_vcf_plus_normal_snv_readcounts> <normal_indel_bam_readcount_file> <DNA|RNA> \
-s <normal_sample_name> -t indel -o <annotated_vcf_plus_normal_readcounts>
```
`annotated_vcf_plus_normal_readcounts` is your final VCF with snv and indel variants and readcounts for all variants in both the tumor and the normal sample.

**Comment by susannasiebert** [comment]
> can it be caused by I copy the wildtype.pm to vep dir manully instead of using pvacseq install ?
No, that is not the issue. This problem occurs if a position in your reference has a different nucleotide base than the Ensembl transcript expects at this position. Since using the `--fasta` input doesn't help here you would need to manually fix the reference base of this variant in the VCF to match the transcript's base at this position and then reannotate with VEP. Most users end up just deleting the variants with this error from their VCF. This should only impact a small fraction of your variants. If it impacts all of your variants this would indicate a bigger problem and we would need to dig deeper.

**Comment by 2236529177** [comment]
thanks a lot.
I am sure my version of vep is compatible.
I also tried the genome offered by vep and genome that I used to generate vcf when annotate with vep, both raised the same error as https://github.com/griffithlab/pVACtools/issues/368
<br>
I also tried to delete the error variant , but then another error variant comes out, so this is not good solution.

I have send my vcf to your email

thanks a lot

**Comment by 2236529177** [comment]
I tried many vcfs, almost each will raise error 
**ERROR: There was a mismatch between the actual wildtype amino acid sequence (S) and the expected amino acid sequence (P). Did you use the same reference build version for VEP that you used for creating the VCF?**

only the test.vcf you supplied in example data it is fine

really spiritless 
thanksa  lot

**Comment by malachig** [comment]
Dear @2236529177,

I think it might be helpful to step back a bit.

Can you summarize briefly:
- What is the nature of your project?
- What kind of raw sequence data are you starting with and where did it come from?
- How did you align your reads?  Against what reference genome?
- How did you call your variants?  
- How did you annotate your variants?  What version of VEP are you using?  What version of the Ensembl VEP cache?

**Comment by 2236529177** [comment]
I also met another problem in my last step. 
command like following

#   deal   tumor   first
python3   bam_readcount_helper.py      decomposed.vcf      tumor   human_g1k_v37.fasta   tumor.bam   NOPREFIX      test   20

then   normal   
python3   bam_readcount_helper.py      decomposed.vcf      normal   human_g1k_v37.fasta   normal.bam   NOPREFIX      test   20

#   add   tumor   annotation
vcf-readcount-annotator   decomposed.vcf   tumor_bam_readcount_snv.tsv   DNA   -s   tumor   -t   snv   -o   tumor_snv_annotated.vcf
vcf-readcount-annotator   tumor_snv_annotated.vcf   tumor_bam_readcount_indel.tsv   DNA   -s   tumor   -t   indel   -o   tumor_annotated.vcf

#   add   normal   annotation
vcf-readcount-annotator   tumor_annotated.vcf   normal_bam_readcount_snv.tsv   DNA   -s   normal   -t   snv   -o   normal_snv_annotated.vcf
vcf-readcount-annotator   normal_snv_annotated.vcf   normal_bam_readcount_indel.tsv      DNA   -s   normal   -t   indel   -o   normal_annotated.vcf

<br>
giving warning for each line, can you tell me where is not fit. thanks  a lot
![image](https://user-images.githubusercontent.com/25199946/101649898-587bea80-3a76-11eb-8302-40cf056c6b21.png)

**Comment by susannasiebert** [comment]
> Warning: variant base -A is not present in the bam-readcount entry for variant 2 171911712. This might indicate that the bam-readcount file doesn't match the VCF. 

If you see these error for the normal sample only, that is to be expected since the normal sample will likely not have the mutations that are tracked here. Do you see these errors also for the tumor sample?

**Comment by 2236529177** [comment]
> > Warning: variant base -A is not present in the bam-readcount entry for variant 2 171911712. This might indicate that the bam-readcount file doesn't match the VCF.
> 
> If you see these error for the normal sample only, that is to be expected since the normal sample will likely not have the mutations that are tracked here. Do you see these errors also for the tumor sample?

only in normal sample, I said it just happened in the last step.
I deal with tumor first ,and then normal
no, here I am asking two questions.
1 is the process I do right?
2 somatic mutaion just happen in tumor sample. why we need the normal sample depth here?

**Comment by 2236529177** [comment]
> Dear @2236529177,
> 
> I think it might be helpful to step back a bit.
> 
> Can you summarize briefly:
> 
> * What is the nature of your project?
> * What kind of raw sequence data are you starting with and where did it come from?
> * How did you align your reads?  Against what reference genome?
> * How did you call your variants?
> * How did you annotate your variants?  What version of VEP are you using?  What version of the Ensembl VEP cache?

thanks a lot

wes or target sequence sample
I used **varcan** to call vcf, 
I used **bwa** to align reads, 
used human_g1k_v37.fasta as reference, 
used vep release 99
used vep to annotate

**I have sent the vcf I used and command used to the email help@pvactools.org**, have you received it?

**Comment by malachig** [comment]
And what does your VEP annotate command l ook like?

**Comment by 2236529177** [comment]
docker run  --rm --user  `id -u`:`id -g` -v /:/data ensemblorg/ensembl-vep vep  --input_file /data/phased.vcf.gz  --output_file /data/test_single/test.vep.vcf --dir_cache /data/database/vep_data_release_99/  --dir_plugins /data/database/vep_data_release_99/VEP_plugins-release-99 --symbol  --fasta /data//database/human_g1k_v37.fasta  --offline  --cache  --transcript_version  --plugin Downstream  --plugin Wildtype  --force_overwrite --vcf --variant_class   --transcript_version   --refseq --assembly GRCh37 --format vcf --cache_version 99 --keep_csq --use_given_ref --tsl --terms SO  --vcf --pick


**# I also tried to replace human_g1k_v37.fasta as vep genome Homo_sapiens.GRCh37.dna.primary_assembly.fa, but keep the same error**, the command I revised some concrete directory as simplified directory to prevent privacy. thanks a lot
docker run -it --rm --user  `id -u`:`id -g` -v /:/data  griffithlab/pvactools   pvacseq run  /data/test_single/test.vep.vcf   tumor  HLA-A*02:01,HLA-B*35:01,DRB1*11:01  MHCflurry NetMHCpan NetMHCIIpan /data/test_single/  -e 8,9,10  --iedb-install-directory /opt/iedb -t 40  --normal-sample-name normal

**Comment by susannasiebert** [comment]
> I have sent the vcf I used and command used to the email help@pvactools.org, have you received it?

Yes, I have received your VCF. I have not had time yet to look into it.

**Comment by 2236529177** [comment]
> > I have sent the vcf I used and command used to the email [help@pvactools.org](mailto:help@pvactools.org), have you received it?
> 
> Yes, I have received your VCF. I have not had time yet to look into it.

ok, thanks a lot.  I am trying different solutions too

**Comment by 2236529177** [comment]
Here I want to give a extremely important issue.
This time I used the **vep Homo_sapiens.GRCh37.dna.primary_assembly.reordered.fa**(the downloaded chromosome order is not right, so I reorder it) to use as reference to call variant by varscan, and I use the vep to annotate with the same release, but it really surprised me, the same error haeppens, I really can  not understand this. because I think vep transcripts cache created by vep genome, hers is Homo_sapiens.GRCh37.dna.primary_assembly.reordered.fa.  so why here still reports 
<br>
**ERROR: There was a mismatch between the actual wildtype amino acid sequence (L) and the expected amino acid sequence (V). Did you use the same reference build version for VEP that you used for creating the VCF?**
OrderedDict([('chromosome_name', '2'), ('start', '38298202'), ('stop', '38298203'), ('reference', 'C'), ('variant', 'G'), ('gene_name', 'CYP1B1'), ('transcript_name', 'NM_000104.3'), ('transcript_support_level', 'NA'), ('amino_acid_change', 'V/L'), ('codon_change', 'Gtg/Ctg'), ('ensembl_gene_id', '1545'), ('hgvsc', ''), ('hgvsp', ''), ('wildtype_amino_acid_sequence', 'MGTSLSPNDPWPLNPLSIQQTTLLLLLSVLATVHVGQRLLRQRRRQLRSAPPGPFAWPLIGNAAAVGQAAHLSFARLARRYGDVFQIRLGSCPIVVLNGERAIHQALVQQGSAFADRPAFASFRVVSGGRSMAFGHYSEHWKVQRRAAHSMMRNFFTRQPRSRQVLEGHVLSEARELVALLVRGSADGAFLDPRPLTVVAVANVMSAVCFGCRYSHDDPEFRELLSHNEEFGRTVGAGSLVDVMPWLQYFPNPVRTVFREFEQLNRNFSNFILDKFLRHCESLRPGAAPRDMMDAFILSAEKKAAGDSHGGGARLDLENVPATITDIFGASQDTLSTALQWLLLLFTRYPDVQTRVQAELDQVVGRDRLPCMGDQPNLPYVLAFLYEAMRFSSFVPVTIPHATTANTSVLGYHIPKDTVVFVNQWSVNHDPLKWPNPENFDPARFLDKDGLINKDLTSRVMIFSVGKRRCIGEELSKMQLFLFISILAHQCDFRANPNEPAKMNFSYGLTIKPKSFKVNVTLRESMELLDSAVQNLQAKETCQ'), ('downstream_amino_acid_sequence', ''), ('fusion_amino_acid_sequence', ''), ('variant_type', 'missense'), ('protein_position', '432'), ('transcript_expression', 'NA'), ('gene_expression', 'NA'), ('normal_depth', 'NA'), ('normal_vaf', 'NA'), ('tdna_depth', '0'), ('tdna_vaf', '0.0'), ('trna_depth', 'NA'), ('trna_vaf', 'NA'), ('index', '37.CYP1B1.NM_000104.3.missense.432V/L'), ('protein_length_change', '')])
<br>

**Comment by 2236529177** [comment]
I have sent a new email to you, it contains more detailed information of the error.

the theme of the email is pVACtools/issues/611, thanks a lot, thanks ma lot

**Comment by malachig** [comment]
So, I know that Ensembl VEP now support the use case where your alignments
are against build37 but you wish to annotate against a transcript
annotation set that was developed on build38 (i.e. Ensembl v99 was built on
build38).  But we haven't really tested this.  I wonder if that is the
source of some issues.

Is there a particular reason why your are not aligning against build38 of
the human genome which has now been out for almost 10 years?

Also it looks like you are using the RefSeq transcript set instead of
Ensembl transcripts.  We also never do that...

**Comment by 2236529177** [comment]
thanks a lot.
vep 99 not only builded in grch38

you can see here
ftp://ftp.ensembl.org/pub/grch37/release-99/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.dna.primary_assembly.fa.gz.


here I have another new error.
runtime/cgo: runtime/cgo: pthread_create failed: Resource temporarily unavailable.
have you ever met this?

**Comment by malachig** [comment]
Those are just the build37 FASTA files.  The existence of these files in Ensembl 99 does NOT imply much about how the *transcript annotations* are created by Ensembl.  Maybe they do now produce a complete transcript annotation build against build37.  At the very least this seems to introduce some limitations.  For example TSL values are not available for build37.  We use this information to prioritize pVACtools results.  Apris values are not available for build37.  This will influencing transcript ranking by VEP and presumably impacts the behavior of --pick and --flag-pick.   "[The gene annotation on the GRCh37 archive is based on Ensembl data from release 75 and will not be updated.](https://uswest.ensembl.org/info/website/tutorials/grch37.html)". Not sure if that is actually still correct, but if true its a bit of a limitation.  Gencode version used for b37 is Gencode19 (current version is 35). I'm even less sure about the RefSeq set used for build37.  Etc., Etc., Etc.

Also, RefSeq transcript models and Ensembl transcript variant models have an important difference in the way the nucleotide sequence is represented.  RefSeq is based largely on full length cDNA sequencing (and the sequences observed in those cDNAs).  Ensembl combines information from multiple sources but extracts the transcript sequence from the reference genome fasta itself.  So maybe RefSeq annotations would be more likely to encounter errors of this type `ERROR: There was a mismatch between the actual wildtype amino acid sequence (S) and the expected amino acid sequence`.  I would still expect them to be fairly rare though... 

In any case, this thread now contains so many issues and confusions that it is hard to offer much more than general advice.  Generally these github issues work best when you report a single, specific, reproducible error.

**Comment by 2236529177** [comment]
thanks a lot
so which version of vep do you suggest, I want to keep the same with you in grch37

**Comment by malachig** [comment]
To be the most in line with what we are doing right now I think you would align your reads against build GRCh38 and annotate your variants using the Ensembl transcript set.  So you would drop `--refseq` and  `--assembly GRCh37`

But that would involve realigning your data.  You might start by first just dropping `--refseq` and see if that eliminates the `ERROR: There was a mismatch between the actual wildtype amino acid sequence (S) and the expected amino acid sequence` issue.

**Comment by worker000000** [comment]
thanks a lot, it is hard to drop grch37 at present.

**Comment by susannasiebert** [comment]
From your last email it sounded like you were able to resolve this issue by adjusting your VEP command. I'm closing this issue.
